{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bcb395e1",
   "metadata": {},
   "source": [
    "Exportação das tabelas com todas as colunas para o banco de dados, para trabalhar outros insigts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2cdb4038",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import psycopg2 as pg\n",
    "from sqlalchemy import create_engine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f994ce74",
   "metadata": {},
   "outputs": [
    {
     "ename": "OperationalError",
     "evalue": "connection to server at \"localhost\" (127.0.0.1), port 5432 failed: Connection refused\n\tIs the server running on that host and accepting TCP/IP connections?\n",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mOperationalError\u001b[39m                          Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 5\u001b[39m\n\u001b[32m      2\u001b[39m arquivos_csv = [f \u001b[38;5;28;01mfor\u001b[39;00m f \u001b[38;5;129;01min\u001b[39;00m os.listdir(pasta_arquivos) \u001b[38;5;28;01mif\u001b[39;00m f.endswith(\u001b[33m'\u001b[39m\u001b[33m.csv\u001b[39m\u001b[33m'\u001b[39m)]\n\u001b[32m      4\u001b[39m engine = create_engine(\u001b[33m'\u001b[39m\u001b[33mpostgresql+psycopg2://postgres:postgres@localhost:5432/comercio_exterior\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m conn = \u001b[43mpg\u001b[49m\u001b[43m.\u001b[49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mdbname=comercio_exterior user=postgres password=postgres host=localhost\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      6\u001b[39m cur = conn.cursor()\n\u001b[32m      8\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m arquivo \u001b[38;5;129;01min\u001b[39;00m arquivos_csv:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.venv/lib/python3.12/site-packages/psycopg2/__init__.py:122\u001b[39m, in \u001b[36mconnect\u001b[39m\u001b[34m(dsn, connection_factory, cursor_factory, **kwargs)\u001b[39m\n\u001b[32m    119\u001b[39m     kwasync[\u001b[33m'\u001b[39m\u001b[33masync_\u001b[39m\u001b[33m'\u001b[39m] = kwargs.pop(\u001b[33m'\u001b[39m\u001b[33masync_\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m    121\u001b[39m dsn = _ext.make_dsn(dsn, **kwargs)\n\u001b[32m--> \u001b[39m\u001b[32m122\u001b[39m conn = \u001b[43m_connect\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdsn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconnection_factory\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconnection_factory\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwasync\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    123\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m cursor_factory \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    124\u001b[39m     conn.cursor_factory = cursor_factory\n",
      "\u001b[31mOperationalError\u001b[39m: connection to server at \"localhost\" (127.0.0.1), port 5432 failed: Connection refused\n\tIs the server running on that host and accepting TCP/IP connections?\n"
     ]
    }
   ],
   "source": [
    "pasta_arquivos = '../../../datalake/gold'\n",
    "arquivos_csv = [f for f in os.listdir(pasta_arquivos) if f.endswith('.csv')]\n",
    "\n",
    "engine = create_engine('postgresql+psycopg2://postgres:postgres@localhost:5432/comercio_exterior')\n",
    "conn = pg.connect(\"dbname=comercio_exterior user=postgres password=postgres host=localhost\")\n",
    "cur = conn.cursor()\n",
    "\n",
    "for arquivo in arquivos_csv:\n",
    "    caminho_completo = os.path.join(pasta_arquivos, arquivo)\n",
    "    nome_tabela = os.path.splitext(arquivo)[0]\n",
    "\n",
    "    print(f\"\\nEnviando '{arquivo}' para a tabela '{nome_tabela}'...\")\n",
    "\n",
    "    try:\n",
    "        df_sample = pd.read_csv(caminho_completo, encoding='latin1', nrows=100)\n",
    "\n",
    "        df_sample = df_sample.astype(str)\n",
    "\n",
    "        df_sample.head(0).to_sql(name=nome_tabela, con=engine, if_exists='replace', index=False)\n",
    "\n",
    "        with open(caminho_completo, 'r', encoding='latin1') as f:\n",
    "            cur.copy_expert(f\"COPY {nome_tabela} FROM STDIN WITH CSV HEADER\", f)\n",
    "\n",
    "        conn.commit()\n",
    "        print(f\"Tabela '{nome_tabela}' carregada com sucesso.\")\n",
    "    except Exception as e:\n",
    "        conn.rollback()\n",
    "        print(f\"Erro ao carregar '{arquivo}': {repr(e)}\")\n",
    "\n",
    "cur.close()\n",
    "conn.close()\n",
    "print(\"\\n Todos os arquivos foram processados.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
